refactor(router): compute correct history restoration when navigation is cancelled

aahmedayed opened the pull request:
We can’t determine whether the user actually meant the `back` or
the `forward` using the popstate event (triggered by a browser
back/forward)
so we instead need to store information on the state and compute the
distance the user is traveling withing the browser history.
So by using the `History#go` method,
we can bring the user back to the page where he is supposed to be after
performing the action.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

atscott commented on packages/common/src/location/location.ts:
```diff
@@ -188,6 +188,16 @@ export class Location {
   }
 
+  // TODO: rename this method to go
+  /**
+   * Navigates in the platform's history.
+   *
+   * @param position  Position of the target page in the history .
+   */
+  goTo(position: number): void {
```
```suggestion
  /**
   * Navigate to a specific page from session history, identified by its relative position to the current page.
   *
   * @param relativePosition  Position of the target page in the history relative to the current page.
   * @see https://developer.mozilla.org/en-US/docs/Web/API/History_API#Moving_to_a_specific_point_in_history
   */
  goTo(relativePosition = 0): void {
```

atscott commented on packages/common/testing/src/location_mock.ts:
```diff
@@ -123,6 +123,16 @@ export class SpyLocation implements Location {
     }
   }
+
+  goTo(position: number): void {
```
```suggestion
  goTo(relativePosition = 0): void {
```

aahmedayed replied:
Hi @atscott - thank you for your review, and yes i think it’s a good idea to have the solution into two PR’s.
I’ve already started to work on it.

atscott commented on packages/common/src/location/location_strategy.ts:
```diff
@@ -162,4 +163,8 @@ export class PathLocationStrategy extends LocationStrategy {
     this._platformLocation.back();
   }
+
+  go(position: number): void {
```
```suggestion
  go(relativePosition = 0): void {
```

atscott commented on packages/common/src/location/platform_location.ts:
```diff
@@ -179,6 +181,10 @@ export class BrowserPlatformLocation extends PlatformLocation {
     this._history.back();
   }
 
+  go(position: number): void {
```
```suggestion
  go(relativePosition = 0): void {
```

atscott reviewed (changes requested):
Hi @aahmedayed - this looks promising! What do you think of breaking this up into two PRs? I think a PR to add `go`/`goTo` as a feature separately would be valuable on its own. Then have a separate PR to fix the routing history with guards. This would make the review process much easier.

I haven't reviewed the changes to the router yet, as that will take a little more time and careful review.

aahmedayed commented:
Hi @atscott - thank you for your review, and yes i think it’s a good idea to have the solution into two PR’s.
I’ve already started to work on it.

aahmedayed commented:
@atscott - i creat a new feature PR #38890 for the go/goTo methodes

atscott commented on packages/router/src/router.ts:
```diff
@@ -672,10 +672,20 @@ export class Router {
                          eventsSubject.next(navCancel);
+                         // Must Reset the navigationId to the lastSuccessfulId when the transition
+                         // navigation was canceled.
+                         this.navigationId = this.lastSuccessfulId;
```
Separate note from the one above:
I'm not super confident that modifying the `navigationId` in this way is a good idea, nor the restoration logic in the `scheduleNavigation`. It's supposed to be a unique id per navigation. 

While this likely won't affect people too much, I can think of one scenario off the top of my head where this could be a problem. The `navigationId` is exposed in the router events so if you have a redirect or cancelled navigation, you're going to be seeing multiple separate navigations using the same id, which isn't super great (and could certainly break tests that are doing assertions on the router events).

 It might be better to use a different field rather than modify the meaning of this one, which is exposed outside the router in various ways.

aahmedayed replied:
Yes @atscott , you were right, i should’ve used `pageId` instead of `navigationId`, so we’ll need to add the `pageId ` field.

aahmedayed committed:
refactor(router): compute correct history restoration when navigation is cancelled

We can’t determine whether the user actually meant the `back` or
the `forward` using the popstate event (triggered by a browser
back/forward)
so we instead need to store information on the state and compute the
distance the user is traveling withing the browser history.
So by using the `History#go` method,
we can bring the user back to the page where he is supposed to be after
performing the action.

implementation for #13586

aahmedayed commented:
Hi @atscott - I’ve already finished the adjustments.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1306,9 +1338,10 @@ export class Router {
       // TODO(jasonaden): Remove first `navigationId` and rely on `ng` namespace.
-      this.location.replaceState(path, '', {...state, navigationId: id});
+      this.location.replaceState(
+          path, '', {...state, navigationId: id, currentPageId: this.currentPageId});
```
nit: looks like this can be extracted to a variable and used for both `location.replaceState` and `location.go` `{...state, navigationId: id, currentPageId: this.currentPageId}`

atscott commented on packages/router/src/router.ts:
```diff
@@ -201,7 +201,7 @@ function defaultMalformedUriErrorHandler(
 
 export type RestoredState = {
-  [k: string]: any; navigationId: number;
+  [k: string]: any; navigationId: number; currentPageId: number;
```
To limit the impact of this breaking change, it might be a good idea to name this something that's unlikely to have collisions. Maybe `ngRouterPageId`

atscott commented on packages/router/src/router.ts:
```diff
@@ -353,6 +352,8 @@ export class Router {
    */
   private lastLocationChangeInfo: LocationChangeInfo|null = null;
   private navigationId: number = 0;
+  private currentPageId: number = 0;
+  private lastSuccessFulPageId: number = 0;
```
```suggestion
  private lastSuccessfulPageId: number = 0;
```

atscott replied:
```suggestion
  private currentPageId = 0;
  private lastSuccessFulPageId = 0;
```

atscott commented on packages/router/src/router.ts:
```diff
@@ -672,7 +674,15 @@ export class Router {
+                         // so we need to go back to the current url if the guard result is false.
+                         if (t.source !== 'imperative') {
+                           const targetPagePosition =
+                               this.lastSuccessFulPageId - this.currentPageId;
```
```suggestion
                               this.lastSuccessFulPageId - t.targetPageId;
```
I think this could be `t.tagetPageId` instead of `currentPageId`, right? That seems to make more sense gramatically: "Calculate the change in positions based on the difference between lastSuccessful and the target navigation"

atscott commented on packages/router/src/router.ts:
```diff
@@ -353,6 +352,8 @@ export class Router {
    */
   private lastLocationChangeInfo: LocationChangeInfo|null = null;
   private navigationId: number = 0;
+  private currentPageId: number = 0;
```
I think it's a little difficult to wrap one's head around the logic here because of the naming. For the duration of a route transition, `this.currentPageId` and `NavigationTransition#targetPageId` are equal. At the end of the navigation, `lastSuccessfulPageId` is set to the `targetPageId`. 

It might be easier to understand if it was like this:
```
/** 
 * The id of the currently active page in the router.
 * Updated to the transition's target id on a successful navigation.
 * Null if there has been no successful navigation yet. 
*/
currentPageId: number|null = null;
/** The id to use for the the next router transition. Increments on a successful navigation. */
nextPageId = 0;
.....

scheduleNavigation() {
  ....
    if (restoredState && restoredState.ngRouterPageId &&
        (isInitialPage || isNotImperativeNavigationTrigger)) {
      this.nextPageId = restoredState.ngRouterPageId;
    } 
    targetPageId = this.nextPageId;
  ....
}

...

this.navigations.subscribe(
        t => {
          this.navigated = true;
          this.lastSuccessfulId = t.id;
          this.currentPageId = t.targetPageId;
          this.nextPageId++;
```

atscott commented on packages/router/src/router.ts:
```diff
@@ -802,6 +812,11 @@ export class Router {
                          eventsSubject.next(navCancel);
                          t.resolve(false);
                        }
+                       // Must Reset the currentPageId to the lastSuccessfulId when the transition
```
I think this reset won't be necessary if you update the logic to what was suggested above by using `this.nextPageId` and `this.currentPageId` and updating them only on a successful navigation rather than in the `scheduleNavigation` function.

atscott reviewed (changes requested):
This is looking better :) I've added another round of comments.

I think you may have missed one of my comments about a request to add some more testing around the history in some scenarios:

* An error during navigation (this was why we moved some logic to `finalize`)
* a route being cancelled by a canLoad guard
* a route being cancelled by a resolver returning `EMPTY`

atscott commented:
Adding the `breaking change` flag since this would break tests that do assertions on the `state` object.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1278,8 +1294,24 @@ export class Router {
+    const isInitialPage = this.currentPageId === 0;
+    // Try to get the current state when page refreshed.
+    if (isInitialPage) {
+      restoredState = this.location.getState() as RestoredState | null;
```
Oh, I almost forgot -- in `setUpLocationChangeListener`, the state is grabbed and the `navigationId` is deleted from the object. I think you should do the same for the new property on `RestoredState` as well.

Also, maybe it would be a good idea to remove the `restoredState` parameter from the `scheduleNavigation` function signature. Since you're now grabbing it here anyways, there's no reason to be passing it in.

atscott replied:
This change actually brought up an issue. While it probably wouldn't really affect things in production, it can break tests because you can do something like
```
router.navigate('xyz');
router.navigate('abc');
location.back();
tick();
```
In production there's really a tick after every navigation but in tests you control timeout flushes. The result is that, because of the `setTimeout` around the navigation scheduled by the `setupLocationChangeListener`, the state is going to be off when this function finally gets triggered. 

I think the best thing to do would be to revert the changes which removed `restoredState` from the parameters for `scheduleNavigation` and instead pass in the state from `navigateByUrl` if `isInitialPage` is true there (rather than doing it here). That way, this function can always respect the value for `restoredState` rather than respecting it all the time _except_ for initial the page load.

aahmedayed commented on packages/router/src/router.ts:
```diff
@@ -270,24 +270,23 @@ export type Navigation = {
-  targetRouterState: RouterState|null,
-  guards: Checks,
-  guardsResult: boolean|UrlTree|null,
+  id: number, targetPageId: number; currentUrlTree: UrlTree,
```
@atscott - I can't fix it, beacause it's the result of `clang-format`.

atscott replied:
Sorry, I should have made the suggested change more minimal and let the formatter do the rest. If you change the semicolon to a comma, this will format correctly.

aahmedayed replied:
Thanks for the clarifications @atscott. I'll fix it.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1278,8 +1294,24 @@ export class Router {
+        (isInitialPage || isNotImperativeNavigationTrigger)) {
+      targetPageId = this.currentPageId = restoredState.currentPageId;
+    } else {
+      targetPageId = ++this.currentPageId;
```
Ah, just thought of another scenario where this could be problematic: if the `NavigationExtras` contains `skipLocationChange`, I think it's going to break the page id tracking since a successful navigation doesn't result in a new item in the browser history.

aahmedayed commented:
Hi @atscott - I've made the adjustments required.

atscott commented on packages/router/src/router.ts:
```diff
@@ -709,7 +718,14 @@ export class Router {
+                           // The navigator change the location before triggered the browser event,
+                           // so we need to go back to the current url if the guard result is false.
+                           const targetPagePosition = this.currentPageId! - t.targetPageId;
+                           this.location.goTo(targetPagePosition);
```
Thinking forward a bit to how this fix might be enabled, I don't think we can do this be default. It's going to have to be opt-in behavior or all the workarounds that people have created to fix the problem on their end will break.

aahmedayed replied:
Let me know if you have a good solution in mind. I'm open to all suggestion 👀.

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -2453,7 +2494,254 @@ describe('Integration', () => {
          expect(location.path()).toEqual('/initial');
        })));
   });
+  describe('`restoredState#ngRouterPageId`', () => {
```
In general, I think these tests look relatively good in covering the different types of scenarios that would cause a cancellation. That said, I don't think they are exercising the path which causes the bug to happen. You're triggering imperative navigations to the cancelled navigations -- these aren't the navigations that are broken though. 
You need to first be able to get the blocked url into the history, navigate away, and then navigate back with a `forward`/`back` to the route being blocked. The assertion that needs to be made then is that the history is still in a good state. That is, the router didn't replace the state for the blocked url in history but instead moved back/forward (in the reverse direction) of the failed navigation.
I'm guessing that by doing this, it will bring up issues around the implementation because you're only doing the `goTo` restoration when _guards_ cancel the navigation but not in other scenarios.

aahmedayed replied:
@atscott , I'm not sure I understand this correctly. Can you please provide me with more details ? Thank you.

atscott replied:
Absolutely! So the issue we're addressing with this change is that the router does not handle history restoration correctly when a navigation is cancelled. More specifically, the restoration is only wrong when the navigation was triggered by the browser's forward or back button, _not_ when the navigation was cancelled by an imperative (`router.navigate` / `router.navigateByUrl`) action. 
The tests below which test the paths to cancelled navigations are all triggered by imperative navigations so we already expect these to work. You will actually need to update the guards/resolvers to be configurable by first allowing a navigation so you can get the router into the history, then navigate away, and try to navigate back with `location.back/forward`.

aahmedayed replied:
ah! Got it @atscott thank you.

atscott replied:
Sorry to have to raise this again, but I don't think the tests have addressed the comment. I can't quite remember what they were before so I don't know what change you made, but the `location.back` or `location.forward` navigation is the navigation that needs to get cancelled. Then performing the action again, but having the guard allow the navigation should result in the correct state. This is what the original tests were doing that you uncommented:
1. create some state where you can go back and forward
2. TestBed.inject(MyCanDeactivateGuard).allow = false;
3.  location.back();
4. assert some state (back was cancelled, state should be restored to what it was before)
5. TestBed.inject(MyCanDeactivateGuard).allow = true;
6.  location.back();
7. assert state again (back was not cancelled. State should be 1 previous in history; before this fix, it was 2 previous)

aahmedayed commented on packages/router/src/router.ts:
```diff
@@ -1313,8 +1331,36 @@ export class Router {
+      targetPageId = this.currentPageId! + 1;
+    }
+
+    // We should always set the value of the `currentPageId` to 1 when the position of
+    // `targetPageId` is 1 and the router hasn't a currentPageId . Because the next navigation can
+    // take the same `pageId` and this can cause history problems.
+    const hasCurrentPage = this.currentPageId !== null;
+    if (!hasCurrentPage && targetPageId === 1) {
+      this.currentPageId = 1;
+    }
```
Hi @atscott, I noticed that, wether the first page is cancelled or not, since it's actually  the first page, the current page ID should always  be set to 1, because the navigator will always save that first page in its history.

aahmedayed replied:
Hi @atscott, I noticed that, wether the first page is cancelled or not, since it's actually the first page, the current page ID should always be set to 1, because the navigator will always save that first page in its history.

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -2453,7 +2494,254 @@ describe('Integration', () => {
+         ]);
+
+         router.navigateByUrl('/one', {skipLocationChange: true});
+         advance(fixture);
```
I think the issue with `skipLocationChange` is the following scenario:

1. Navigate to `'/one'` (ngRouterPageId === 1)
2. Navigate to `'/two'` (ngRouterPageId === 2)
3. Navigate to `'/three'` with `skipLocationChange` true (ngRouterPageId === 3)
4. Navigate with `location.back` and have the guard cancel the navigation
 ===> The restoration logic will calculate the difference between the `ngRouterPageId` to be 2 because of `skipLocationChange`, but we're actually only navigating back 1

aahmedayed replied:
I think when we navigate to `/three` with skipLocationChange equals true, the ngRouterPageId should  have the value of 2. Also the navigator will only have 2 histories.

atscott replied:
Ah, yes that's right! I was thinking about this case again over the weekend and forgot why I had previously realized it would be fine (because we skip the browser url update in this case 🤦 )

IgorMinar reviewed (dismissed):
As mentioned in https://github.com/angular/angular/pull/38890#pullrequestreview-501426763 let's figure out if we can fix the targeted issue via a new public api like `historyGo`(or similar). Once the fix is confirmed to work, then we can discuss the public api, although `historyGo` while awkward, seems to be the least choice given the unfortunate precedence set by `history.go`.

Reviewed-for: public-api

aahmedayed commented on packages/router/test/integration.spec.ts:
```diff
@@ -2453,7 +2494,181 @@ describe('Integration', () => {
+
+         router.navigateByUrl('/two');
+         advance(fixture);
+         expect(location.getState()).toEqual(jasmine.objectContaining({ngRouterPageId: 2}));
```
Hi @atscott, See the comment below.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1356,7 +1393,22 @@ export class Router {
+        {navigationId: this.lastSuccessfulId, ngRouterPageId: this.currentPageId});
+  }
+
+  private cancelNavigationTransition(t: NavigationTransition, reason = '') {
```
Rather than making this default to `''`, let's make this required and have callers explicitly pass empty string if necessary.

atscott replied:
It looks like you can also refactor `finalize` of the transitions rxjs pipe/stream to use this function.

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -2727,109 +3033,189 @@ describe('Integration', () => {
+
+             TestBed.inject(MyCanDeactivateGuard).allow = true;
+
+             TestBed.inject(MyCanActivateGuard).allow = false;
+             location.goTo(-2);
+             advance(fixture);
+             expect(location.path()).toEqual('/third');
+             expect(fixture.nativeElement).toHaveText('child3');
+             expect(location.getState()).toEqual(jasmine.objectContaining({ngRouterPageId: 4}));
```
```suggestion

```
I think this is a duplicate test to the one section above. Please remove

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -2727,109 +3033,189 @@ describe('Integration', () => {
+
+             TestBed.inject(MyCanDeactivateGuard).allow = true;
+
+             TestBed.inject(MyCanActivateGuard).allow = false;
+             location.goTo(2);
+             advance(fixture);
+             expect(location.path()).toEqual('/third');
+             expect(fixture.nativeElement).toHaveText('child3');
+             expect(location.getState()).toEqual(jasmine.objectContaining({ngRouterPageId: 4}));
```
```suggestion

```
I think this is a duplicate test to the one section above. Please remove

atscott reviewed (changes requested):
Nice work so far. I think this is at the point where some testing in existing applications needs to be done. I don't see anything more that jumps out as concerning.

atscott commented:
Hi @aahmedayed, just wanted to ping and check if you're still working on this PR. My last set of comments (specifically https://github.com/angular/angular/pull/38884/files#r505020217 and https://github.com/angular/angular/pull/38884/files#r505932975) didn't appear to be addressed in your Nov 24 push, which is why I haven't reviewed the changes again.
Also, with regards to https://github.com/angular/angular/pull/38884#discussion_r498927057, you can address this by making this new behavior configurable in the `ExtraOptions` (https://angular.io/api/router/ExtraOptions)

aahmedayed commented:
Hi @atscott , i apologize for not following-up with the matter recently, i’ve had some personal stuff to deal with, i’m back now.
Quick update: i’ve made some improvements to the existing solution and completed the changes that you requested.

aahmedayed commented:
Hi @atscott  , i apologize for not following-up with the matter recently, i’ve had some personal stuff to deal with, i’m back now.
Quick update: i’ve made some improvements to the existing solution and completed the changes that you requested.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1391,7 +1406,22 @@ export class Router {
+  }
+
+  private cancelNavigationTransition(t: NavigationTransition, reason: string) {
+    if (this.canceledNavigationResolution === 'corrected' && t.source === 'popstate') {
```
What about `urlUpdateStrategy: 'eager'`? If we're navigating imperatively with `urlUpdateStrategy: 'eager'`, shouldn't the restoration strategy be to go back rather than replace state?

atscott commented on packages/router/src/router_module.ts:
```diff
@@ -427,6 +427,17 @@ export interface ExtraOptions {
+   *
+   * The default value is `legacy`
+   */
+  canceledNavigationResolution?: 'legacy'|'corrected';
```
maybe we should make this `'replace'|'computed'`? The old behavior was to always perform `'replaceState'` regardless of what type of navigation happened. The new behavior is more of a "computed" restoration where we try to determine the correct thing to do.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1391,7 +1406,22 @@ export class Router {
     this.location.replaceState(
-        this.urlSerializer.serialize(this.rawUrlTree), '', {navigationId: this.lastSuccessfulId});
+        this.urlSerializer.serialize(this.rawUrlTree), '',
+        {navigationId: this.lastSuccessfulId, ngRouterPageId: this.currentPageId});
```
I think we maybe shouldn't add the `ngRouterPageId` to the state object if the option isn't enabled. Otherwise we might still need to consider this a breaking change.

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -5592,10 +5911,10 @@ describe('Integration', () => {
            router.events.subscribe(e => e instanceof RouterEvent && events.push(e));
 
-           location.go('/include/user/kate(aux:excluded)');
+           router.navigateByUrl('/include/user/kate(aux:excluded)');
```
It looks like you have a bunch of changes to the tests as a result of your update to the location mock. That's a bit concerning and certainly indicates that would be a breaking change, even if it's more correct than what the mock currently does.

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -5557,14 +5876,14 @@ describe('Integration', () => {
 
            // another unsupported URL
-           location.go('/exclude/two');
+           router.navigateByUrl('/exclude/two');
```
why was this changed?

atscott reviewed (changes requested):
This is starting to look really good! The rewritten tests are great. I have a couple notes and some concerns around the changes to the location mock and the subscription in the router. I'll need to do some more testing on this change and give it another pass.

aahmedayed commented:
Hi @atscott  i made the requested change.
Also, i created a new commit e3da2c6eb3 but i'm not sure if we should add it in a new Pr?

atscott commented:
@aahmedayed Thanks, I'll take another look today. I think we should not include e3da2c6 because it would make [this TODO](https://github.com/angular/angular/blob/5c9bd237e34c38ae03a4defd5776aba56deec131/packages/common/src/location/location.ts#L92-L93) even harder to resolve. It might be that it's something that doesn't make sense to change anymore but we should limit the changes in this PR to what's absolutely necessary to accomplish our goal here.

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -2662,7 +2662,256 @@ describe('Integration', () => {
+         advance(fixture);
+         expect(location.getState()).toEqual(jasmine.objectContaining({ngRouterPageId: 3}));
+       }));
+
```
We're missing handling for `replaceUrl` when setting the target page id. The following test fails:
```
    fit('should work when using `NavigationExtras.replaceUrl`', fakeAsync(() => {
          const location = TestBed.inject(Location);
          const router = TestBed.inject(Router);

          expect(location.path()).toEqual('/second');
          expect(location.getState()).toEqual(jasmine.objectContaining({ngRouterPageId: 3}));

          router.navigateByUrl('/first', {replaceUrl: true});
          advance(fixture);
          expect(location.getState()).toEqual(jasmine.objectContaining({ngRouterPageId: 3}));
          expect(location.path()).toEqual('/first');
        }));
```

atscott commented on packages/common/testing/src/location_mock.ts:
```diff
@@ -91,7 +91,7 @@ export class SpyLocation implements Location {
     const url = path + (query.length > 0 ? ('?' + query) : '');
     this.urlChanges.push(url);
-    this._subject.emit({'url': url, 'pop': false});
+    this._subject.emit({'url': url, 'pop': false, 'type': 'imperative'});
```
Does this change need to be made? In reality, I think this would trigger a `'popstate'` in the browser if you were to use the real `Location.go`, would it not?

aahmedayed replied:
Yes @atscott you're right, we not need this change, and for your second point, the browser will not trigger a popstate for `history.push` and `history.replace` methods.

atscott commented on packages/router/test/integration.spec.ts:
```diff
@@ -2662,7 +2662,256 @@ describe('Integration', () => {
+        {
+          path: 'first',
+          component: SimpleCmp,
+          canActivate: [MyCanActivateGuard],
+          resolve: [MyResolve]
+        },
+        {
+          path: 'second',
+          component: SimpleCmp,
+          canDeactivate: [MyCanDeactivateGuard],
+          canActivate: [MyCanActivateGuard]
+        },
+        {
+          path: 'third',
+          component: SimpleCmp,
+          canDeactivate: [MyCanDeactivateGuard],
```
So that it's easier to understand below, I would suggest adding both guards to each path rather than having to remember which one has which guard.

atscott reviewed (changes requested):
Another round of comments with one noticeable bug. 

Sorry, I know this has been a ton of work for you but you're doing a really great job here! This change is incredibly complex with a lot of moving parts and edge cases to test.

aahmedayed commented:
Hi @atscott, I'll try to finish with the requested changes this week at the latest.

aahmedayed commented:
Hi @atscott, I'm done with the requested changes, but i think we can't do this change https://github.com/angular/angular/pull/38884#discussion_r627654160 now because the test_ivy_aot failed and the error is that cannot find name 'RestoredState' in `router.d.ts`.

atscott reviewed (changes requested):
@aahmedayed thanks for the updates! I'll take a look at the changes this week. No problem on the `@internal` comment. 

My plan for landing this change is:
* Submit the change first without a public API option so we can continue to test and offer the solution to the interested parties on the bug as a private preview (the option in the Router will be set to `@internal` but users can opt-in with `(Router as any).canceledNavigationResolution = 'computed'`). We can consider a follow-up change to add it as public in a minor version.
* Move the changes to `location_mock` into a local copy used by the router tests. This avoids making this a breaking change and we can remove it when #41730 is submitted.
* clean up the commits: squash most/all of them into a single commit and update the commit message(s)
* Move the commit that updated the `navigationId` and `pageId` to be in the `ng` namespace to a follow-up PR (again to avoid making this a breaking change). I like this change but it would prevent us from submitting this until v13.

Sorry for requesting reverts to changes I had you do; we should get them merged eventually, but getting this PR submitted sooner rather than later (in v13) seems like a good idea. Do you mind if I make those changes and push up to your branch as a co-author or would you like to do the updates yourself?

aahmedayed commented:
Ok @atscott there is no probleme, I start making those changes right now. But I'm not sure I understand the second point correctly. Can you please provide me with more details ? Thank you.

atscott commented:
@aahmedayed - Absolutely. There are changes here that depend on your updates to the back/forward methods in `SpyLocation`. Those changes will be considered a breaking change and can't be submitted until v13. To get around that, we can copy the whole implementation of `SpyLocation` to a class in the router's `integration.spec.ts` and make the update there (and also add `{provide: Location, useClass: RouterSpyLocation}` to the providers in the `beforeEach`). That way we can submit this change now and then later update the router test to point to the real `SpyLocation` in v13 after #38884 is submitted.

aahmedayed commented:
Ah ok! I understand, thank you @atscott

aahmedayed commented:
Hi @atscott - I'm done with the requested changes.

atscott commented on packages/router/src/events.ts:
```diff
@@ -93,7 +93,8 @@ export class NavigationStart extends RouterEvent {
       navigationTrigger: 'imperative'|'popstate'|'hashchange' = 'imperative',
       /** @docsNotRequired */
-      restoredState: {[k: string]: any, navigationId: number}|null = null) {
+      restoredState: {[k: string]: any, navigationId: number, ngRouterPageId?: number}|
```
Maybe this was missed in one of the squashes. Let's remove `ngRouterPageId` here so we don't change the public API.

atscott commented on packages/router/src/router.ts:
```diff
@@ -626,9 +642,11 @@ export class Router {
+                                     ...t.extras.state,
+                                     navigationId: t.id,
+                                     ngRouterPageId: t.targetPageId
+                                   });
```
Unresolving this comment as well. `ngRouterPageId` should only be set if `canceledNavigationResolution === 'computed'`. This change was probably reverted when you added then had to remove the `ng` property on state.

atscott replied:
Also, the idea for `generateNgRouterState` was a bit lost, I think because you also wanted to use it in `resetUrlToCurrentUrlTree`. I still think it's confusing to add and then delete the property. Can we change the function to be 
```
private generateNgRouterState(navigationId: number,  ngRouterPageId?: number, state?: [key: string]: any) {
  const state = {...(state ?? {})};
  state.navigationId = navigationId;
  if (this.canceledNavigationResolution === 'computed') {
    state.ngRouterPageId = ngRouterPageId;
    };
  }
  return state;
}
```
Then this can be used by both `setBrowerUrl` (you'll have to revert the change to that function so you keep the original `state` separate from the additional router information) as well as in `resetUrlToCurrentUrlTree`.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1390,8 +1433,38 @@ export class Router {
+    if (this.canceledNavigationResolution === 'replace') {
+      delete state.ngRouterPageId;
+    }
+    return state;
```
This should be `{...(state ?? {})}` or the behavior of `setBrowserUrl` will change to no longer make a copy of the state.

atscott commented on goldens/size-tracking/integration-payloads.json:
```diff
@@ -38,10 +38,10 @@
-        "polyfills-es2015": 36975,
-        "src_app_lazy_lazy_module_ts-es2015": 825
+        "runtime-es2015": 2854,
+        "main-es2015": 296439,
```
This seems a little suspicious given that `main-es2015` is showing double the size from before. Can you revert the change to this file, rebase the PR on master, and then see what CI says the number should be?

atscott reviewed (changes requested):
Looking good. A few more minor comments. Also, can you amend the commit message to be:

```
refactor(router): compute correct history restoration when navigation is cancelled

We can’t determine whether the user actually meant the `back` or
the `forward` using the popstate event (triggered by a browser back/forward)
so we instead need to store information on the state and compute the 
distance the user is traveling withing the browser history.
So by using the `History#go` method,
we can bring the user back to the page where he is supposed to be after
performing the action.
```

We should change the scope of this commit to be a refactor since we haven't provided a public API for the fix (so nothing was actually fixed yet). We should also remove the `fix #` comment since, again, we haven't made the fix available.

aahmedayed commented:
Hi @atscott - I'm done with the requested changes.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1391,7 +1431,37 @@ export class Router {
+      // When navigation gets cancelled while using url update strategy eager, then we have to use
+      // `location#back` method. Because, when `urlUpdateSrategy` is `eager`; `setBrowserUrl` method
+      // is called before any verification.
+      else if (this.urlUpdateStrategy === 'eager') {
```
Can't this case be combined with the one above: `t.source === 'popstate' || this.urlUpdateStrategy === 'eager'`? The computed `historyGo` should be `-1` at that point, which would be equivalent to `this.location.back()`.

atscott commented on packages/router/src/router.ts:
```diff
@@ -1390,8 +1423,34 @@ export class Router {
+        this.location.historyGo(targetPagePosition);
+      } else if (this.urlUpdateStrategy === 'eager') {
+        this.location.back();
+      }
```
Unresolving. Please add the `else` case and add a comment as to why nothing is done there (if update is not `'eager'`, then the navigation was cancelled before any browser url change so nothing needs to be restored).

atscott reviewed (approved):
LGTM with two minor comments on the state restoration function.

reviewed-for: public-api
reviewed-for: size-tracking

AndrewKushnir commented:
[Presubmit](http://test/OCL:377944420:BASE:377944453:1623085980672:2741e6d).

AndrewKushnir commented on packages/router/src/router.ts:
```diff
@@ -501,6 +510,22 @@ export class Router {
   relativeLinkResolution: 'legacy'|'corrected' = 'corrected';
 
+  /**
+   * TODO: Add this API option as public in a minor version
+   * @internal
```
Could you please move these two lines to the end of the JSDoc block? That'd be more consistent with other JSDoc annotations.

AndrewKushnir commented on packages/router/src/router.ts:
```diff
@@ -501,6 +510,22 @@ export class Router {
+   * The default value is `replace`
+   *
+   */
+  canceledNavigationResolution: 'replace'|'computed' = 'replace';
```
Thanks for adding docs 👍

I'm still a bit confused about the difference from an API user's perspective: can we add a bit more info on how the behavior changes depending on selected option? I guess we can also consider updating option names to reflect that as it looks like these options reflect the inner functionality more than observed behavior change from user perspective.

atscott replied:
The behavior is meant to be an implementation detail rather than completely transparent. In fact, it shouldn’t be an option at all but may need to be in order to allow migration without a breaking change. Still tbd what this will look like when made into public api (or if we forgo the option and release as a breaking change bug fix in v13). I get that the existing computed vs replace was a good balance between explaining the difference and keeping details hidden. Do you have another suggestion for the options?

AndrewKushnir replied:
That makes sense too, I was confused by the TODO above it (that says "TODO: Add this API option as public in a minor version"). I think the solution might be to update the TODO using the information from your comment and mention why this field was introduced and what possible scenarios are.

jessicajaniuk reviewed (approved):
LGTM 🍪 
reviewed-for: size-tracking

AndrewKushnir commented on packages/router/src/router.ts:
```diff
@@ -626,9 +652,10 @@ export class Router {
-                                   this.setBrowserUrl(
-                                       t.urlAfterRedirects, !!t.extras.replaceUrl, t.id,
-                                       t.extras.state);
+                                   this.setBrowserUrl(t.urlAfterRedirects, !!t.extras.replaceUrl, {
+                                     ...t.extras.state,
+                                     ...this.generateNgRouterState(t.id, t.targetPageId)
+                                   });
```
It looks like some of the code is repeated here and below, we can probably refactor it a bit to move common code to the private function (like `setBrowserUrlWithState`), for example:
```suggestion
                                   this.setBrowserUrlWithState(t.urlAfterRedirects, t);
```
Similar to what you did with extracting common logic to the `cancelNavigationTransition` function (which is great!).

AndrewKushnir commented on packages/router/src/router.ts:
```diff
@@ -1348,8 +1378,20 @@ export class Router {
     }
 
     const id = ++this.navigationId;
+    let targetPageId: number = 0;
```
I'd propose adding a bit more context here (as a comment in the code) and describe what `0` represents.

AndrewKushnir commented on packages/router/src/router.ts:
```diff
@@ -1391,7 +1431,37 @@ export class Router {
+        this.generateNgRouterState(this.lastSuccessfulId, this.currentPageId));
+  }
+
+  private cancelNavigationTransition(t: NavigationTransition, reason: string) {
```
It'd be great to add some docs for this function since it performs a number of operations, so having more context would be great (btw, thanks for adding comments inside the function!).

atscott commented on packages/router/src/router.ts:
```diff
@@ -233,7 +233,9 @@ function defaultMalformedUriErrorHandler(
-  [k: string]: any; navigationId: number;
+  [k: string]: any,
+  navigationId: number,
+  ngRouterPageId?: number,
```
The `ng` prefix is there to reduce the chance of colliding with any existing user properties on the history state. We eventually want to move both of these under and `ng` property anyways to avoid this issue in the future but can't do that until v13. https://github.com/angular/angular/pull/38884#discussion_r627641640

AndrewKushnir replied:
Thanks for additional context, that makes sense! I think we may consider using `ɵ` as a prefix, i.e. `ɵrouterPageId` as the `ɵ` symbol is known to to Angular-internal symbol (like static `ɵcmp` in Ivy). I hope there are no browser quirks around using it in `window.location` API :) I'd also propose adding this info as a comment, so it's available for additional context during future refactoring.

aahmedayed commented:
Hi guys, i have finished the changes that have been asked of me, but i haven’t had the chance to work on the docs since i’m having a very busy week.
@atscott if you don’t mind, could you please take it from here to fix the documentations so that i don’t cause a delay in the workflow. Thank you in advance.

AndrewKushnir reviewed (approved):
@aahmedayed thanks for addressing the feedback, the changes look good 👍

atscott commented:
another [presubmit](http://test/OCL:337182271:BASE:378521808:1623279962617:aee1229f) for a sanity check

aahmedayed commented:
You're the most welcome @AndrewKushnir and thanks a lot @atscott

atscott commented:
Hi @aahmedayed, I had a thought tonight about this and realized we missed a pretty important scenario. It's very common for guards to schedule a new navigation before returning `false`. This could be a bit of an issue since `history.go` causes a `popstate` event which will schedule another navigation that will compete with the one triggered by the guard. What we might want to do in the `finalize`/`cancelNavigationTransition` is to treat this scenario (`!completed && !errored`) more like what's done when a guard returns a `UrlTree`. This logic is [here](https://github.com/angular/angular/blob/983c540191b5d838e19863c1e722170be6ea42e3/packages/router/src/router.ts#L894-L929) and if you look through it, you'll see that it's treated as a redirect and `resetStateAndUrl` is not called.

A specific scenario where this doesn't work:
1. `urlUpdateStrategy: 'eager'`
2. navigate to `/a`
3. navigate to `/b`
4. click back to go to `/a`, which has a `canActivate` guard that navigates to `/`
 ===> expectation: end on `/` but we end on `/b` instead because of the `history.go` call that triggers a navigation.

There might be other scenarios I haven't thought of. I'll be on vacation until July. If you get a chance to take a look and send a PR before then, I can review it when I'm back. Otherwise I'll see about investigating other scenarios, making some adjustments to the code, and then start thinking about making the option public opt-in.

atscott commented:
@aahmedayed I've been working with this more today and I'm a little concerned that this won't work in the end. We can always figure out how to handle things when navigations go through the back/forward buttons or through `router.navigate`. However, navigations from outside of that can't be handled well (`Location.go`, any regular `href` on the page, browser navigation with the navbar, AngularJS/Angular hybrid apps where navigations come from AngularJS). All of these situations could result in something worse than what's already in place because these won't have the correct `routerPageId` and the computation for `history.go` will be incorrect.

aahmedayed commented:
HI @atscott - Sorry for the delay I’m taking some days off, I’ll be back Monday. Thank you for all the fixes you made.
